# Using CellTypist Models on Spatial Data
*Created by Jo Adams from the Single-cell Analytics Innovation Labratory (SAIL) at MSKCC with code from Andrew Moorman and Emma Chen* 

This repository contains examples on how to train and run CellTypist Models on 10X Xenium Spatial Transcriptomics data.

There are four explanatory notebooks: <br/>
[**`Training Xenium CellTypist Model Using Xenium`**](https://github.com/joadams1/spatial_celltypist/blob/main/notebooks/Training%20Xenium%20CellTypist%20Model%20Using%20Xenium.ipynb) provides an explanation on how to cell type Xenium data using marker genes and then how to train a CellTypist model on that data <br/>
[**`Training Spatial CellTypist Model using scRNAseq`**](https://github.com/joadams1/spatial_celltypist/blob/main/notebooks/Training%20Spatial%20CellTypist%20Model%20using%20scRNAseq.ipynb) similarly provides an explanation on how to use an annotated single-cell RNA-seq dataset to train a CellTypist model <br/>
[**`Using Xenium CellTypist Model`**](https://github.com/joadams1/spatial_celltypist/blob/main/notebooks/Using%20Xenium%20CellTypist%20Model.ipynb) walks through how to apply a spatial CellTypist Model to a new Xenium dataset <br/>
[**`Basic Analysis of Xenium Data`**](https://github.com/joadams1/spatial_celltypist/blob/main/notebooks/Basic%20Analysis%20of%20Xenium%20Data.ipynb) shows how to do some basic analysis and visualization for a cell typed Xenium dataset. 

For more information on CellTypist and an in-depth explanation of how it works on scRNA-seq data, refer to this earlier [SAIL GitHub](https://github.com/joadams1/celltypist). If you have any questions, please reach out to the [computational SAIL team](https://sail.mskcc.org/research/computational-research/).  
