# Using CellTypist Models on Spatial Data
*Created by Jo Adams with code from Andrew Moorman from the Single-cell Analytics Innovation Labratory (SAIL) at MSKCC* 

This repository contains examples on how to train and run CellTypist Models on 10X Xenium Spatial Transcriptomics data.

There are 3 explanatory notebooks: <br/>
[**Training Xenium CellTypist Model**](https://github.com/joadams1/spatial_celltypist/blob/main/Training%20Xenium%20CellTypist%20Model.ipynb) provides an explanation on how to cell type Xenium data using marker genes and then how to train a CellTypist model on that data <br/>
[**Using Xenium CellTypist Model**](https://github.com/joadams1/spatial_celltypist/blob/main/Using%20Xenium%20CellTypist%20Model.ipynb) walks through how to apply a CellTypist Model trained on Xenium data to a new Xenium dataset <br/>
[**Basic Analysis of Xenium Data**](https://github.com/joadams1/spatial_celltypist/blob/main/Basic%20Analysis%20of%20Xenium%20Data.ipynb) shows how to do some basic analysis and visualization for a cell typed Xenium dataset. 

For more information on CellTypist and an in-depth explanation of how it works on scRNA-seq data, refer to this earlier [SAIL GitHub](https://github.com/joadams1/celltypist). If you have any questions, please reach out to the [computational SAIL team](https://sail.mskcc.org/research/computational-research/).  
